cgi-bin/DW/BlobStore.pm

```diff
@@ -0,0 +1,157 @@
+        unless $namespace =~ m!^(?:[a-z][a-z0-9]+)$!;
+}
+
+sub ensure_key_is_valid {
+    my ( $key ) = @_;
+
+    # This is just a check to ensure that nobody uses a key without path
+    # elements or with invalid characters.
+    $log->logcroak( "Key '$key' is invalid." )
+        unless $key =~ m!^(?:[a-z0-9]+[_:/-])+([a-z0-9]+)$!;
+    return 1;
+}
+
+sub store {
+    my ( $class, $namespace, $key, $blobref ) = @_;
+    ensure_namespace_is_valid( $namespace );
+    ensure_key_is_valid( $key );
+    $log->logcroak( 'Store requires data be a scalar reference.' )
+        unless ref $blobref eq 'SCALAR';
+    $log->debug( "Meta-blobstore: storing ($namespace, $key)" );
+
+    # Enforce read-only mode
+    if ( $LJ::DISABLE_MEDIA_UPLOADS ) {
+        $log->info( 'Denying write due to $LJ::DISABLE_MEDIA_UPLOADS being set.' );
+        return 0;
+    }
+
+    # Storage requests always go to the first blobstore that will take them,
+    # we never store something twice.
+    foreach my $bs ( @{$class->_get_blobstores} ) {
+        my $rv = $bs->store( $namespace, $key, $blobref );
+        return $rv if $rv;
+    }
+    $log->info( "Meta-blobstore: failed to store ($namespace, $key)" );
+    return 0;
+}
+
+sub delete {
+    my ( $class, $namespace, $key ) = @_;
+    ensure_namespace_is_valid( $namespace );
+    ensure_key_is_valid( $key );
+    $log->debug( "Meta-blobstore: deleting ($namespace, $key)" );
+
+    # Enforce read-only mode
+    if ( $LJ::DISABLE_MEDIA_UPLOADS ) {
+        $log->info( 'Denying write due to $LJ::DISABLE_MEDIA_UPLOADS being set.' );
+        return 0;
+    }
+
+    # Deletes must be sent to all blobstores. Return true if any accepted
+    # the delete.
+    my $rv = 0;
+    foreach my $bs ( @{$class->_get_blobstores} ) {
+        $rv = $bs->delete( $namespace, $key ) || $rv;
+    }
+    return $rv;
+}
+
+# Retrieves a file from the blobstore. May return either a scalar-ref if the file
+# was found, or returns undef.
+sub retrieve {
+    my ( $class, $namespace, $key ) = @_;
+    ensure_namespace_is_valid( $namespace );
+    ensure_key_is_valid( $key );
+    $log->debug( "Meta-blobstore: retrieving ($namespace, $key)" );
+
+    # Try blobstores in priority order.
+    foreach my $bs ( @{$class->_get_blobstores} ) {
+        my $rv = $bs->retrieve( $namespace, $key );
+        return $rv if $rv;
+    }
+    $log->info( "Meta-blobstore: failed to retrieve ($namespace, $key)" );
+    return undef;
+}
+
+sub exists {
+    my ( $class, $namespace, $key ) = @_;
+    ensure_namespace_is_valid( $namespace );
+    ensure_key_is_valid( $key );
+    $log->debug( "Meta-blobstore: checking if exists ($namespace, $key)" );
+
+    # Try blobstores in priority order.
+    foreach my $bs ( @{$class->_get_blobstores} ) {
+        my $rv = $bs->exists( $namespace, $key );
+        return $rv if $rv;
+    }
+    $log->info( "Meta-blobstore: file doesn't exist in any store ($namespace, $key)" );
+    return undef;
+}
+
+1;
```

Comment:
I'm assuming that you've tested all the functionality of this and the various submodules (except S3 which looks like a placeholder), but I'd also like to see some automated tests for these methods.